src/transformers/models/donut/modeling_donut_swin.py

```diff
@@ -531,7 +531,7 @@ def __init__(self, config, dim, input_resolution, num_heads, shift_size=0):
         self.output = DonutSwinOutput(config, dim)
 
     def set_shift_and_window_size(self, input_resolution):
-        if min(input_resolution) <= self.window_size:
+        if min(input_resolution) < self.window_size:
```

Comment:
Any reason this gets updated?

Reply:
This change was required in modeling_swin.py to resolve the issue of mismatch in outputs we were facing with checkpoints of mask2former using Swin-B backbone. 
Not sure if modeling_donut_swin.py also needs the same fix. @alaradirik let me know if this change needs to be reverted

Reply:
Can you confirm the slow integration tests of models like Donut and Swin Transformer are not affected by this change?
```
RUN_SLOW=yes pytest tests/models/swin
```

Reply:
sure, I'll check this

Reply:
It might be that we'll need to leverage `MaskFormerSwinBackbone` for Mask2Former as well.

We cannot merge the PR if the slow integration tests are failing

Reply:
In HuggingFace we just took this code from timm: https://github.com/rwightman/pytorch-image-models/blob/9f5bba9ef9db8a32a5a04325c8eb181c9f13a9b2/timm/models/swin_transformer.py#L252-L256

Reply:
Update: OneFormer author informed me that MaskFormer, Mask2Former and OneFormer all use the same backbone, so it should work with our SwinBackbone

Reply:
This was only for Swin-base checkpoints?

Reply:
> This line overwrites the shift size of the final layer of the final transformer block as 0 (should be 6). Since it's specific to the base model, it's not effecting OneFormer.

Thanks for the explanation, @alaradirik! I am curious about what's different in the base model, though. The downsampling layers output the same resolution for all Swin variants, right? Also, is this issue for the cityscapes base model or the coco base model, or both? Thanks

Reply:
No problem @praeclarumjj3! @shivalikasingh95 can correct me if I'm wrong but it is specific to coco-base. SWIN variants' downsampling layers output the same resolution but the shift and window sizes are different, which triggers the overwriting method

Reply:
small correction here - This issue is not specific to coco-base. This issue exists for Swin-B, Swin-B (IN21k) and Swin-L variants.

We were initially using coco checkpoints with small, tiny and base variants for our conversion and so we encountered the issue with the base checkpoint. 
But I double checked last night for models corresponding to other datasets and the issue is observed for all of them i.e. with the ones using Swin-B, Swin-B (IN21k) and Swin-L as backbones.

Reply:
> Thank you!
> 
> I'm just wondering why the issue was occurring only on Swin-base on one specific dataset. It would definitely be nice to clear that up, does it have to do with the image resolution?
> 
> For instance for UperNet (at #20648) I was able to perfectly convert all checkpoints that leverage Swin-base by using our `SwinBackbone`. This one was ported from the mmsegmentation library whose Swin implementation is [here](https://github.com/open-mmlab/mmsegmentation/blob/master/mmseg/models/backbones/swin.py#L166). So it's a bit strange. Might it be that we were just "lucky" with UperNet and OneFormer?

@NielsRogge I debugged this too last night. I figured out why the UperNet and OneFormer don't come across this issue but why we ran into this issue.

It's because we are using MaskFormerImageProcessor which is resizing our input image to (384,384) where as for OneFormer input is being resized to (640,640) and UperNet is making use of SegFormerImageProcessor which is resizing input to (512,512).

So with pixel_values of shape [1, 3, 384, 384] as input to the backbone, we end up with following dimensions in downsampling layers:
(96, 96)
(48, 48)
(24, 24)
(12, 12)

With 12 as `window_size` (used by Swin-B, Swin-B (IN21k) and Swin-L of Mask2Former) and `input_resolution` as (12,12), the method below overwrites the shift_size:

```
def set_shift_and_window_size(self, input_resolution):
        if min(input_resolution) <= self.window_size:
            # if window size is larger than input resolution, we don't partition windows
            self.shift_size = 0
            self.window_size = min(input_resolution)

```

Hope this clears up the issue.

Reply:
Oh thanks a lot for clearing that up!

That's very valuable. So it depends on the initial image resolution and window size, that's interesting.

Reply:
Glad you found it helpful! This turned out to be a great learning exercise for me :)

Although I'm just curious why [mmseg's swin implementation](https://github.com/open-mmlab/mmsegmentation/blob/master/mmseg/models/backbones/swin.py#L166) doesn't include the above logic for overwriting shift and window size based on input resolution because the [original swin implementation](https://github.com/microsoft/Swin-Transformer/blob/main/models/swin_transformer.py#L206) includes it.

Mask2Former [even MaskFormer & OneFormer] backbone implementation is based on [this](https://github.com/SwinTransformer/Swin-Transformer-Semantic-Segmentation/blob/main/mmseg/models/backbones/swin_transformer.py) which also seems to be taken from mmseg. 

May be there's a simple explanation here but since this is all still new for me...it's possible that I'm missing some information.
I checked original MaskFormer for any clues regarding this small tweak in the backbone but didn't find anything. 

Let me know in case you guys have any idea on this @NielsRogge @alaradirik @praeclarumjj3

Reply:
Hmm, thanks for looking into this, @shivalikasingh95! The Swin authors changed the implementation for their semantic segmentation repo, as that's the one that everyone followed, including mmseg, detectron2, etc.

I suppose the original authors of [Swin's Semantic Segmentation repo](https://github.com/SwinTransformer/Swin-Transformer-Semantic-Segmentation/blob/main/mmseg/models/backbones/swin_transformer.py) replaced the overwriting method with a [few lines of code that pad](https://github.com/facebookresearch/Mask2Former/blob/9b0651c6c1d5b3af2e6da0589b719c514ec0d69a/mask2former/modeling/backbone/swin.py#L250-L255) the input features to a multiple of window size (ensuring `window_partition` is always applicable) to support arbitrary resolution inference. 
```python
# pad feature maps to multiples of window size
pad_l = pad_t = 0
pad_r = (self.window_size - W % self.window_size) % self.window_size
pad_b = (self.window_size - H % self.window_size) % self.window_size
x = F.pad(x, (0, 0, pad_l, pad_r, pad_t, pad_b))
_, Hp, Wp, _ = x.shape
```

Consequently, the mmseg, MaskFormer, Mask2Former, OneFormer, and other segmentation frameworks follow the same padding functionality (is this used in HF transformers @NielsRogge?). So, I don't think it's accurate to add this to the `SwinBackbone` (as it's only meant for segmentation frameworks that all follow the padding-based implementation). 

There must be some differences between the original segmentation-focused implementation and SwinBackbone, though, that lead to the inconsistency of backbone features. @NielsRogge might be able to clear it up further.

Let me know if there are any corrections to my reasoning. Thanks a lot to everyone for looking into this. I was not aware of this odd inconsistency.

Reply:
`SwinBackbone` modifies the Swin Transformer encoder used for image classification by 1) outputting hidden states before downsampling, and 2) applying a layernorm to each of the relevant hidden states.

Our `SwinModel` + `SwinBackbone` do support any image resolution (as long as both the height and width are divisible by 32) by dynamically padding. However, to be fully compliant with the original implementation, `SwinBackbone` should not include [this line](https://github.com/huggingface/transformers/blob/bd9d51263a92a5d109811af0d72469f6f8f74aba/src/transformers/models/swin/modeling_swin.py#L648) when using `SwinLayer` . Correct me if I'm wrong, but that would make it possible for Mask2Former to use `SwinBackbone` rather than `MaskFormerSwinBackbone`.

If that's the case, then there are 2 options:

1. add an argument to the forward of `SwinLayer`, `SwinStage` and `SwinEncoder` to specify whether or not to run that line. `SwinBackbone` could then use `SwinEncoder` and specify to not run that line. However that slightly goes against our policy of not tweaking files (one model, one file) and `modeling_swin.py` is mainly about Swin for image classification. This would make modeling_swin.py implement both Swin for image classification and Swin as backbone.
2. keep using `MaskFormerSwinBackbone`

Reply:
Option 1 sounds more elegant to me if it **does not break anything else in the library**.

Reply:
> The Swin authors changed the implementation for their semantic segmentation repo, as that's the one that everyone followed, including mmseg, detectron2, etc.
> 
> I suppose the original authors of [Swin's Semantic Segmentation repo](https://github.com/SwinTransformer/Swin-Transformer-Semantic-Segmentation/blob/main/mmseg/models/backbones/swin_transformer.py) replaced the `window_size` overwriting method with a [few lines of code that pad](https://github.com/facebookresearch/Mask2Former/blob/9b0651c6c1d5b3af2e6da0589b719c514ec0d69a/mask2former/modeling/backbone/swin.py#L250-L255) the input features to a multiple of window size (ensuring `window_partition` is always applicable with a constant window size) to support arbitrary resolution inference.

> Consequently, the mmseg, MaskFormer, Mask2Former, OneFormer, and other segmentation frameworks follow the same padding functionality.

Thanks a lot for the above explanation @praeclarumjj3. That clarifies my doubt. 

> However, to be fully compliant with the original implementation, SwinBackbone should not include [this line](https://github.com/huggingface/transformers/blob/bd9d51263a92a5d109811af0d72469f6f8f74aba/src/transformers/models/swin/modeling_swin.py#L648) when using SwinLayer . Correct me if I'm wrong, but that would make it possible for Mask2Former to use SwinBackbone rather than MaskFormerSwinBackbone.

@NielsRogge yes, that's correct. If that line is removed or run based on an argument then we can use `SwinBackbone` instead of `MaskFormerSwinBackbone` for Mask2Former.

> If that's the case, then there are 2 options:
> 1. add an argument to the forward of SwinLayer, SwinStage and SwinEncoder to specify whether or not to run that line. SwinBackbone could then use SwinEncoder and specify to not run that line. However that slightly goes against our policy of not updating files several months after (at the time, modeling_swin.py was implemented to support Swin for image classification). This would make modeling_swin.py implement both Swin for image classification and Swin as a fully compliant backbone.
> 2. keep using MaskFormerSwinBackbone

Based on the points discussed in this thread and my understanding of `SwinBackbone` and `MaskFormerSwinBackbone` implementation...another option is coming to mind. 

Would it make sense to move all the `SwinBackbone` implementation to a different file ?
That way `modeling_swin.py` remains unaffected and it's purpose can still be what it was originally designed for -> i.e. to support Swin for image classification.
We can create a separate file for e.g `backbone_swin.py` and let this be the main file responsible for implementing Swin as a fully compliant backbone. That way if any other changes are needed in future also w.r.t using Swin as backbone then those changes will affect only this file and not the modeling file.

Either way whether we use above option or option 1...I think `MaskFormerSwinBackbone` can be removed from transformers because if running of [this line](https://github.com/huggingface/transformers/blob/bd9d51263a92a5d109811af0d72469f6f8f74aba/src/transformers/models/swin/modeling_swin.py#L648) is made optional then `SwinBackbone` and `MaskFormerSwinBackbone` essentially become identical in functionality. (Feel free to correct me, if that's not the case.)

So in that case even `MaskFormer` should work with `SwinBackbone` too. However, it would require to update all MaskFormer checkpoints on the Hub because there's a minor difference in names of layers when using `SwinBackbone` vs `MaskFormerSwinBackbone`.

This is what I'm thinking based on my understanding but I guess you guys would have much better idea on this @NielsRogge and @alaradirik

Reply:
Thanks for your message @shivalikasingh95!

My current idea would be to go for option 1, i.e. add a boolean argument to determine whether or not to run that line. This way, Mask2Former and OneFormer will be able to use `SwinBackbone`. 

`MaskFormerSwinBackbone` can stay in the library for backwards compatibility with the MaskFormer checkpoints, but we can encourage users to use `SwinBackbone` if they prefer to train a MaskFormer model from scratch.

I'll add this boolean argument in the UperNet PR, which I'm planning to merge either tomorrow or on Monday.

Reply:
@NielsRogge the above sounds good to me. Let me know once UperNet PR is merged. Will update this PR too to switch back to using `SwinBackbone`.

Reply:
UperNet has been merged, so you can proceed with using `SwinBackbone` :)

sorry for the need to switch here again, I know converting checkpoints is a lot of work

Reply:
@NielsRogge @shivalikasingh95 I switched backed to using SwinBackbone but I will need to update all uploaded models.